pennylane/devices/qubit/apply_operation.py

```diff
@@ -0,0 +1,179 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""appling operation to a statevector."""
+from functools import singledispatch
+from string import ascii_letters as ABC
+
+import pennylane as qml
+
+from pennylane import math
+
+SQRT2INV = 1 / math.sqrt(2)
+
+
+def apply_operation_einsum(op: qml.operation.Operator, state):
+    """Apply ``Operator`` to ``state`` using ``matheinsum``. This is more efficent at lower qubit
+    numbers.
+
+    Args:
+        op (Operator): Operator to apply to the
+        state (array[complex]): input state
+
+    Returns:
+        array[complex]: output_state
+    """
+    mat = op.matrix()
+
+    total_indices = len(state.shape)
+    num_indices = len(op.wires)
+
+    state_indices = ABC[:total_indices]
+    affected_indices = "".join(ABC[i] for i in op.wires)
+
+    new_indices = ABC[total_indices : total_indices + num_indices]
+
+    new_state_indices = state_indices
+    for old, new in zip(affected_indices, new_indices):
+        new_state_indices = new_state_indices.replace(old, new)
+
+    einsum_indices = f"{new_indices}{affected_indices},{state_indices}->{new_state_indices}"
+
+    reshaped_mat = math.reshape(mat, [2] * (num_indices * 2))
+
+    return math.einsum(einsum_indices, reshaped_mat, state)
+
+
+def apply_operation_tensordot(op: qml.operation.Operator, state):
+    """Apply ``Operator`` to ``state`` using ``math.tensordot``. This is more efficent at higher qubit
+    numbers.
+
+    Args:
+        op (Operator): Operator to apply to the
+        state (array[complex]): input state
+
+    Returns:
+        array[complex]: output_state
+    """
+    mat = op.matrix()
+    total_indices = len(state.shape)
+    num_indices = len(op.wires)
+    reshaped_mat = math.reshape(mat, [2] * (num_indices * 2))
+    axes = (tuple(range(num_indices, 2 * num_indices)), tuple(op.wires))
+
+    tdot = math.tensordot(reshaped_mat, state, axes=axes)
+
+    unused_idxs = [i for i in range(total_indices) if i not in op.wires]
+    perm = list(op.wires) + unused_idxs
+    inv_perm = math.argsort(perm)
+
+    return math.transpose(tdot, inv_perm)
+
+
+@singledispatch
+def apply_operation(op: qml.operation.Operator, state):
+    """Apply and operator to a given state.
+
+    Args:
+        op (Operator): The operation to apply to ``state``
+        state (ndarray): The starting state.
+
+    Returns:
+        ndarray: output state
+
+    .. warning::
+
+        ``apply_operation`` is an internal function, and thus subject to change without a deprecation cycle.
+
+    .. warning::
+        ``apply_operation`` applies no validation to its inputs.
+
+        This function assumes that the wires of the operator correspond to indices
+        of the state. See :func:`~.map_wires` to convert operations to integer wire labels.
```

Comment:
this comment is great to have, but I'm curious what the next step might be for this. Will we pre-validate/process wires in the future? I guess since the state won't really _know_ if it has some non-int/unordered wire labels, it shouldn't affect this, so we could just convert `op.wires` to `range(len(op.wires))` if needed. I am ok with the responsibility remaining with the calling context (again, wanna make sure that's considered)

Reply:
So these functions will be used inside `PythonDevice.execute`.  `PythonDevice.preprocess` will convert all the wire labels to "indices".

This way we don't have to carry around a wire map every where.